File: C/RicercaBinaria.c
/*Ricerca Binaria su un array di valori*/

#include <stdio.h>
#include <stdlib.h>

int RicercaBinaria(int, int *, int, int);

int main(){
    int x = 23;
    int sx = 0;
    int dx = 9;
    int a[] = {2,3,5,6,8,9,10,12,23,41};

    if(RicercaBinaria(x, a, sx, dx)){
        printf("Il valore %d esiste nella lista.\n",x);
    }
    else
    {
        printf("Il valore %d non esiste nella lista.\n",x);
    }
}

int RicercaBinaria(int x, int *a, int sx, int dx){
    int cx = 0;
    int ans;
    cx = (dx - sx)/2 + sx;
    printf("sx: %d cx: %d dx: %d\n",sx,cx,dx);

    if(a[cx] == x){
        return a[cx];
    }

    if(a[cx] < x && sx <= dx){
        sx = cx;
        ans = RicercaBinaria(x, a, sx, dx);
        return ans;
    }

    if(a[cx] > x && sx <= dx){
        dx = cx;
        ans = RicercaBinaria(x, a, sx, dx);
        return ans;
    }

    return 0;
}

File: Python/RicercaBinaria.py
# -*- coding: utf-8 -*-
"""
Created on Sun Mar 27 11:08:56 2022

ALGORITMO: Ricerca Binaria

Ricerca k nella lista a, ritorna la posizione di k in a o -1 se a non contiene k.
Precondizione: a è una lista di elementi ordinati dal più piccolo al più grande

@author: fabio
"""

def bin_search(a, k):
    n = len(a)
    lx, rx = 0, n-1
    while lx <= rx: # fintanto che lo spazio di ricerca non è vuoto
        cx = (lx + rx)//2
        if k == a[cx]:
            return cx
        if k < a[cx]:
            rx = cx-1
        else: # k > a[cx]
            lx = cx+1
    return -1


a = [ 1, 1, 2, 2, 5, 5, 5, 10, 10, 12, 12, 13, 15, 16, 16, 20, 22, 22 ]
print(bin_search(a, 40))
print(bin_search(a, 10))